From 363da2316a91ef842bc2e9d6d60c88068668c16d Mon Sep 17 00:00:00 2001
From: Varsha Maira <vmaira@habana.ai>
Date: Sat, 15 Jun 2024 14:39:39 +0300
Subject: [PATCH] [FSW-418] fs1: bootrom: configurable wait for tpm2_init

This patch adds configurable timeouts for the
tpm2_init execution.

With a short timeout, tpm2_init can be invoked
periodically, allowing other ROM tasks to continue
in parallel. Alternatively, a long timeout enables
tpm2_init to execute in a single operation.

Change-Id: If8dc32a34c9b93f5e1173dfbb8d0435c4e5fa903
Signed-off-by: Varsha Maira <vmaira@habana.ai>
---

diff --git a/common/tpm2.c b/common/tpm2.c
index 92495e9..4658643 100644
--- a/common/tpm2.c
+++ b/common/tpm2.c
@@ -509,11 +509,14 @@
 	return ret;
 }
 
-static int tpm_tis_wait_init(int loc)
+static int tpm_tis_wait_init(int loc, bool long_timeout_en)
 {
 	uint64_t start;
 	uint8_t status;
+	uint32_t timeout;
 	int ret;
+
+	timeout = long_timeout_en ? MS_TO_US(TPM_LONG_TIMEOUT_MS) : TPM_SHORT_TIMEOUT_US;
 
 	start = boot_rom_get_ts_us();
 	do {
@@ -524,12 +527,12 @@
 		if ((status & TPM_ACCESS_VALID) && (status != 0xff))
 			return 0;
 
-	} while (!boot_rom_timeout_expired(start, MS_TO_US(TIS_LONG_TIMEOUT_MS)));
+	} while (!boot_rom_timeout_expired(start, timeout));
 
 	return -1;
 }
 
-int tpm2_init(void)
+int tpm2_init(bool long_timeout_en)
 {
 	int ret;
 
@@ -538,7 +541,7 @@
 	if (!tpm_is_supported())
 		return 0;
 
-	ret = tpm_tis_wait_init(locality);
+	ret = tpm_tis_wait_init(locality, long_timeout_en);
 	if (ret)
 		return ret;
 
diff --git a/common/tpm2.h b/common/tpm2.h
index b970d7d..33e8850 100644
--- a/common/tpm2.h
+++ b/common/tpm2.h
@@ -8,7 +8,7 @@
 
 #include <hl_tpm2.h>
 
-int tpm2_init(void);
+int tpm2_init(bool long_timeout_en);
 int tpm2_startup_send_cmd(enum tpm2_startup_types mode);
 int tpm2_startup_recv_resp(void);
 int tpm2_pcr_extend_send_cmd(uint32_t index, uint32_t algorithm,
